### PR TITLE
Channel for Google hangouts chat integration

### PR DESCRIPTION
**Pull request by muelerma**
**Proposed changes**:
- Please see issue #5006 for details

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by erohmensing** [comment]
Thanks for the PR! We'll give it a review as soon as possible.

File: tests/core/test_channels.py [hunk 1]
```diff
@@ -911,3 +911,210 @@ def test_has_user_permission_to_send_messages_to_conversation_without_permission
     assert not RasaChatInput._has_user_permission_to_send_messages_to_conversation(
         jwt, message
     )
+
+
+def test_hangouts_channel():
```
Comment (degiz) [inline tests/core/test_channels.py]: Thanks a lot for adding tests!

One optional thing: could you please move hangouts related tests to tests/core/channels/hangouts.py?
It's one big test file now and it becomes quite difficult to follow.

File: rasa/core/channels/hangouts.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+    def _text_button_card(self, text: Text, buttons: List) -> Dict:
+        hangouts_buttons = []
+        for b in buttons:
+            try:
+                b_txt, b_pl = b["title"], b["payload"]
+            except KeyError:
+                logger.error(
+                    "Buttons must be a list of dicts with 'title' and 'payload' as keys"
+                )
```
Comment (degiz) [inline rasa/core/channels/hangouts.py]: does it make sense to whether `pass` or `return` after Exception?

File: rasa/core/channels/hangouts.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        removed_intent: Optional[Text] = None,
+    ) -> None:
+
+        self.user_added_intent = user_added_intent
+        self.room_added_intent = room_added_intent
+        self.removed_intent = removed_intent
+
+    @classmethod
+    def name(cls) -> Text:
+        return "hangouts"
```
Comment (degiz) [inline rasa/core/channels/hangouts.py]: this name appears twice, maybe we can extract it to constant?

**Comment by muelerma** [comment]
Hello Degiz,
thanks for your comments. I implemented most of the changes, except the credentials one, please see the comment. Additionally, Google is deprecating token based authorization for OAuth2, which I now added to the channel.

File: rasa/core/channels/hangouts.py (hunk 2 above)
Comment (muelerma) [reply rasa/core/channels/hangouts.py]: Added return

File: tests/core/test_channels.py (hunk 1 above)
Comment (muelerma) [reply tests/core/test_channels.py]: Done that

File: rasa/core/channels/hangouts.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
+import asyncio
```
[the remark this answers is not in the document]
Comment (muelerma) [reply rasa/core/channels/hangouts.py]: Cleaned the imports (were initially just copied from channel.py)

File: docs/user-guide/connectors/hangouts.rst [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+send one response. The response will be displayed to the user either as text or a so-called card (for
+more information, see the Cards section).
+
+The possibility to implement asynchronous communication between Hangouts Chat and bot exists, but due
+to the usually synchronous nature of Rasa bots, this functionality is not included in this channel. 
+
+Running On Hangouts Chat
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you want to connect to Facebook using the run script, e.g. using:
```
[the remark this answers is not in the document]
Comment (muelerma) [reply docs/user-guide/connectors/hangouts.rst]: Yes, fixed that.

File: rasa/core/channels/hangouts.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    """
+
+    @classmethod
+    def from_credentials(cls, credentials: Optional[Dict[Text, Any]]) -> InputChannel:
+        credentials = credentials or {}
+        return cls(
+            credentials.get("user_added_intent"),
+            credentials.get("room_added_intent"),
+            credentials.get("removed_intent"),
+        )
```
[the remark this answers is not in the document]
Comment (muelerma) [reply rasa/core/channels/hangouts.py]: Just to make sure I understand this right, I should add the from_configuration() method to InputChannel (why not only HangoutsChannel?) and this method would read a file called "hangouts_config.yml" with the intent specified in it?

File: rasa/core/channels/hangouts.py (hunk 3 above)
Comment (muelerma) [reply rasa/core/channels/hangouts.py]: Here I orientated myself on the other channels like facebook.py, where both InputChannel and OuputChanel need a name() classmethod and these also return the same value.

File: rasa/core/channels/hangouts.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (degiz) [reply rasa/core/channels/hangouts.py]: Sorry for confusion,

I meant that inside HangoutsChannel you can have fields like `hangouts_user_added_intent _name`, and so on.

[the remark this answers is not in the document]
Comment (degiz) [reply rasa/core/channels/hangouts.py]: And this fields will contain actual Strings that you want to use.

File: rasa/core/channels/hangouts.py (hunk 3 above)
Comment (degiz) [reply rasa/core/channels/hangouts.py]: you are right! I just meant that `hangouts` itself could be extracted to a const, and name() will return this constant.

**Comment by tmbo** [comment]
@muelerma we'd love to get this in the next release (next tuesday). do you have time to look at the suggestions from @degiz to make sure we are ready to go?

**Comment by muelerma** [comment]
@tmbo ok, nice to hear, I will work on it over the weekend.

File: rasa/core/channels/hangouts.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (muelerma) [reply rasa/core/channels/hangouts.py]: Hi Degiz, thanks, I think I got it now. You mean the values should rather be hard-coded in the hangouts.py. When you say fields, do you mean attributes of the HangoutsInput class?

File: rasa/core/channels/hangouts.py (hunk 3 above)
Comment (muelerma) [reply rasa/core/channels/hangouts.py]: Done that

File: rasa/core/channels/hangouts.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (muelerma) [reply rasa/core/channels/hangouts.py]: Please see the latest commit. Thanks your support!

**Comment by muelerma** [comment]
> Hey @muelerma
> 
> thanks for answering the comments! I've just requested two more changes, and we'll be ready to merge!
> 
> about the OAuth2: I've tested the channel without it, and it seems to work fine. Google was using the webhook without any auth. Maybe we can just then remove mentions of OAuth from the description?

Hi @degiz 
which description do you mean, the `/docs/user-guide/connectors/hangouts.rst`? Modified that, please have a look. 

Yes, Google just sends the token, it is up to the bot implementation to check it, therefore it also works without it. If we want the token verification to be optional, I think it is good practice to make the `project_id` parameter, which is required for OAuth to work, also optional. That's why I modified `from_credentials` and constructor method of `HangoutsInput`, please let me know what you think.

File: rasa/core/channels/hangouts.py (hunk 3 above)
Comment (degiz) [reply rasa/core/channels/hangouts.py]: that's exactly it, thanks!

File: rasa/core/channels/hangouts.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (degiz) [reply rasa/core/channels/hangouts.py]: thanks, I meant exactly that!

**Pull request closed without merging**